File: test/python/payment_gateway/refund_order_test.py
import os
import pytest
import time
import asyncio
from uuid import uuid4
from dana.utils.snap_configuration import SnapConfiguration, AuthSettings, Env
from dana.payment_gateway.v1.enum import *
from dana.payment_gateway.v1.models import *
from dana.payment_gateway.v1.models import RefundOrderRequest
from dana.payment_gateway.v1 import *
from dana.api_client import ApiClient
from dana.exceptions import *
from helper.api_helpers import get_headers_with_signature, execute_and_assert_api_error
from helper.util import get_request, with_delay, retry_on_inconsistent_request
from helper.assertion import *
from payment_gateway.payment_pg_util import automate_payment_pg
from concurrent.futures import ThreadPoolExecutor

title_case = "RefundOrder"
create_order_title_case = "CreateOrder"
json_path_file = "resource/request/components/PaymentGateway.json"
user_phone_number = "0811742234"
user_pin = "123321"
merchant_id = os.environ.get("MERCHANT_ID", "216620010016033632482")

# Set up configuration for the API client
configuration = SnapConfiguration(
    api_key=AuthSettings(
        PRIVATE_KEY=os.environ.get("PRIVATE_KEY"),
        ORIGIN=os.environ.get("ORIGIN"),
        X_PARTNER_ID=os.environ.get("X_PARTNER_ID"),
        ENV=Env.SANDBOX
    )
)

# Create an instance of the API client
with ApiClient(configuration) as api_client:
    api_instance = PaymentGatewayApi(api_client)

# Function to generate a unique partner reference number
def generate_partner_reference_no():
    return str(uuid4())

# Function to create an order request
def create_test_order_init():
    data_order = []
    """Helper function to create a test order"""
    case_name = "CreateOrderApi"
    
    # Get the request data from the JSON file
    json_dict = get_request(json_path_file, "CreateOrder", "CreateOrderApi")
    
    # Set the partner reference number
    json_dict["partnerReferenceNo"] = generate_partner_reference_no()
    json_dict["merchantId"] = merchant_id

    # Convert the request data to a CreateOrderRequest object
    create_order_request_obj = CreateOrderByApiRequest.from_dict(json_dict)
    
    # Make the API call
    api_response = api_instance.create_order(create_order_request_obj)
    data_order.append(api_response.partner_reference_no)
    data_order.append(api_response.web_redirect_url)
    print(f"Created order with reference number: {api_response.partner_reference_no}")
    print(f"Web redirect URL: {api_response.web_redirect_url}")
    return data_order

@pytest.fixture(scope="module")
def test_order_reference_number():
    """Fixture that creates a test order once per module and shares the reference number"""
    data_order = create_test_order_init()
    print(f"\nCreating shared test order with reference number: {data_order[0]}")
    return data_order[0]

def create_test_order_paid():
    """Helper function to create a test order with short expiration date to have paid status"""
    data_order = create_test_order_init()
    asyncio.run(automate_payment_pg(
        phone_number=user_phone_number,
        pin=user_pin,
        redirectUrlPayment=data_order[1],
        show_log=True
    ))
    return data_order[0]

@with_delay()
def test_refund_order_in_progress(test_order_reference_number):
    """Test refund order in progress"""
    case_name = "RefundOrderInProgress"

    # Get the request data from the JSON file
    json_dict = get_request(json_path_file, title_case, case_name)

    # Set the partner reference number
    json_dict["partnerReferenceNo"] = test_order_reference_number
    json_dict["merchantId"] = merchant_id

    # Convert the request data to a RefundOrderRequest object
    refund_order_request_obj = RefundOrderRequest.from_dict(json_dict)

    # Make the API call and assert the response
    api_response = api_instance.refund_order(refund_order_request_obj)

    # Assert the response
    assert_response(json_path_file, title_case, case_name, RefundOrderResponse.to_json(api_response), {"partnerReferenceNo": json_dict["originalPartnerReferenceNo"]})

@with_delay()
@retry_test(max_retries=3,delay_seconds=10)
def test_refund_order_valid(test_order_reference_number):
    # Create a test order paid and get the reference number
    test_order_reference_number = create_test_order_paid()
    
    """Test refund order in progress"""
    case_name = "RefundOrderValidScenario"

    # Get the request data from the JSON file
    json_dict = get_request(json_path_file, title_case, case_name)

    # Set the partner reference number
    json_dict["originalPartnerReferenceNo"] = test_order_reference_number
    json_dict["partnerRefundNo"] = test_order_reference_number
    json_dict["merchantId"] = merchant_id

    # Convert the request data to a RefundOrderRequest object
    refund_order_request_obj = RefundOrderRequest.from_dict(json_dict)

    # Make the API call and assert the response
    api_response = api_instance.refund_order(refund_order_request_obj)

    # Assert the response
    assert_response(json_path_file, title_case, case_name, RefundOrderResponse.to_json(api_response), {"partnerReferenceNo": json_dict["originalPartnerReferenceNo"]})

@with_delay()
@retry_test(max_retries=3, delay_seconds=10)
@pytest.mark.skip(reason="skipped by request: scenario RefundOrderExceedsTransactionAmountLimit")
def test_refund_order_due_to_exceed():
    # Create a test order paid and get the reference number
    test_order_reference_number = create_test_order_paid()
    
    """Test refund order in progress"""
    case_name = "RefundOrderExceedsTransactionAmountLimit"

    # Get the request data from the JSON file
    json_dict = get_request(json_path_file, title_case, case_name)

    # Set the partner reference number
    json_dict["originalPartnerReferenceNo"] = test_order_reference_number
    json_dict["partnerRefundNo"] = test_order_reference_number
    json_dict["merchantId"] = merchant_id

    # Convert the request data to a RefundOrderRequest object
    refund_order_request_obj = RefundOrderRequest.from_dict(json_dict)

    # Make the API call and assert the response
    try:
        api_instance.refund_order(refund_order_request_obj)

        pytest.fail("Expected ForbiddenException but the API call succeeded")
    except ForbiddenException as e:
        assert_fail_response(json_path_file, title_case, case_name, e.body, {"partnerReferenceNo": json_dict["originalPartnerReferenceNo"]})

@with_delay()
def test_refund_order_not_allowed(test_order_reference_number):
    """Test refund order when refund is not allowed by agreement"""
    case_name = "RefundOrderNotAllowed"
    
    # Get the request data from the JSON file
    json_dict = get_request(json_path_file, title_case, case_name)
    
    # Set the partner reference number
    json_dict["originalPartnerReferenceNo"] = test_order_reference_number
    json_dict["partnerRefundNo"] = test_order_reference_number
    json_dict["merchantId"] = merchant_id
    
    # Convert the request data to a RefundOrderRequest object
    refund_order_request_obj = RefundOrderRequest.from_dict(json_dict)
    
    # Make the API call and assert the response
    try:
        api_instance.refund_order(refund_order_request_obj)

        pytest.fail("Expected ForbiddenException but the API call succeeded")
    except ForbiddenException as e:
        assert_fail_response(json_path_file, title_case, case_name, e.body, {"partnerReferenceNo": json_dict["originalPartnerReferenceNo"]})
    except:
        pytest.fail("Expected ForbiddenException but the API call give another exception")

@with_delay()
def test_refund_order_due_to_exceed_refund_window_time(test_order_reference_number):
    """Test refund order due to exceed refund window time"""
    case_name = "RefundOrderDueToExceedRefundWindowTime"
   
    # Get the request data from the JSON file
    json_dict = get_request(json_path_file, title_case, case_name)
   
    # Set the partner reference number
    json_dict["originalPartnerReferenceNo"] = test_order_reference_number
    json_dict["partnerRefundNo"] = test_order_reference_number
    json_dict["merchantId"] = merchant_id
   
    # Convert the request data to a RefundOrderRequest object
    refund_order_request_obj = RefundOrderRequest.from_dict(json_dict)
   
    # Make the API call and assert the response
    try:
        api_instance.refund_order(refund_order_request_obj)
        pytest.fail("Expected ForbiddenException but the API call succeeded")
    except ForbiddenException as e:
        assert_fail_response(json_path_file, title_case, case_name, e.body, {"partnerReferenceNo": json_dict["originalPartnerReferenceNo"]})
    except:
        pytest.fail("Expected ForbiddenException but the API call give another exception")

@with_delay()
def test_refund_order_multiple_refund(test_order_reference_number):
    """Test refund order multiple refund"""
    case_name = "RefundOrderMultipleRefund"
    json_dict = get_request(json_path_file, title_case, case_name)
    json_dict["originalPartnerReferenceNo"] = test_order_reference_number
    json_dict["partnerRefundNo"] = test_order_reference_number
    json_dict["merchantId"] = merchant_id
    refund_order_request_obj = RefundOrderRequest.from_dict(json_dict)
    try:
        api_instance.refund_order(refund_order_request_obj)
        pytest.fail("Expected ForbiddenException but the API call succeeded")
    except ForbiddenException as e:
        assert_fail_response(json_path_file, title_case, case_name, e.body, {"partnerReferenceNo": json_dict["originalPartnerReferenceNo"]})
    except:
        pytest.fail("Expected ForbiddenException but the API call give another exception")

@with_delay()
def test_refund_order_not_paid(test_order_reference_number):
    """Test refund order not paid"""
    case_name = "RefundOrderNotPaid"
    json_dict = get_request(json_path_file, title_case, case_name)
    json_dict["originalPartnerReferenceNo"] = test_order_reference_number
    json_dict["partnerRefundNo"] = test_order_reference_number
    json_dict["merchantId"] = merchant_id
    
    refund_order_request_obj = RefundOrderRequest.from_dict(json_dict)
    print(f"Refund order request object: {refund_order_request_obj.to_json()}")
    try:
        api_instance.refund_order(refund_order_request_obj)
        pytest.fail("Expected ServiceException but the API call succeeded")
    except NotFoundException as e:
        assert_fail_response(json_path_file, title_case, case_name, e.body, {"partnerReferenceNo": json_dict["originalPartnerReferenceNo"]})
    except:
        pytest.fail("Expected ServiceException but the API call give another exception")

@with_delay()
@retry_test(max_retries=3,delay_seconds=10)
def test_refund_order_duplicate_request():
    """Test refund duplicate request"""
    case_name = "RefundOrderDuplicateRequest"
    
    # Create a test order paid and get the reference number
    test_order_reference_number = create_test_order_paid()

    # Get the request data from the JSON file
    json_dict = get_request(json_path_file, title_case, case_name)

    # Set the partner reference number
    json_dict["originalPartnerReferenceNo"] = test_order_reference_number
    json_dict["partnerRefundNo"] = test_order_reference_number
    json_dict["merchantId"] = merchant_id

    # Convert the request data to a RefundOrderRequest object
    refund_order_request_obj = RefundOrderRequest.from_dict(json_dict)

    # Make the API call and assert the response
    try:
        # First hit API
        api_instance.refund_order(refund_order_request_obj)
        refund_order_request_obj.refund_amount = Money(currency="IDR", value="10000.00")
        # Second hit API with same data (duplicate)
        api_instance.refund_order(refund_order_request_obj)
    except NotFoundException as e:
        assert_fail_response(json_path_file, title_case, case_name, e.body, {"partnerReferenceNo": json_dict["originalPartnerReferenceNo"]})


@with_delay()
def test_refund_order_illegal_parameter(test_order_reference_number):
    """Test refund order illegal parameter"""
    case_name = "RefundOrderIllegalParameter"

    # Get the request data from the JSON file
    json_dict = get_request(json_path_file, title_case, case_name)

    # Set the partner reference number
    json_dict["partnerReferenceNo"] = test_order_reference_number
    json_dict["merchantId"] = merchant_id

    # Convert the request data to a RefundOrderRequest object
    refund_order_request_obj = RefundOrderRequest.from_dict(json_dict)

    # Make the API call and assert the response
    try:
        api_instance.refund_order(refund_order_request_obj)
        pytest.fail("Expected BadRequestException but the API call succeeded")
    except BadRequestException as e:
        assert_fail_response(json_path_file, title_case, case_name, e.body, {"partnerReferenceNo": json_dict["originalPartnerReferenceNo"]})
    except:
        pytest.fail("Expected BadRequestException but the API call give another exception")

@with_delay()
def test_refund_order_invalid_mandatory_field(test_order_reference_number):
    """Test refund order when mandatory field is invalid"""
    # Refund order
    case_name = "RefundOrderInvalidMandatoryParameter"

    # Get the request data from the JSON file
    json_dict = get_request(json_path_file, title_case, case_name)
    
    json_dict["originalPartnerReferenceNo"] = test_order_reference_number
    json_dict["partnerRefundNo"] = test_order_reference_number
    json_dict["merchantId"] = merchant_id

    # Convert the request data to a RefundOrderRequest object
    refund_order_request_obj = RefundOrderRequest.from_dict(json_dict)
    
    headers = get_headers_with_signature(
        method="POST",
        resource_path="/payment-gateway/v1.0/debit/refund.htm",
        request_obj=json_dict,
        with_timestamp=True,
        invalid_signature=None
    )

    execute_and_assert_api_error(
        api_client,
        "POST",
        "http://api.sandbox.dana.id/payment-gateway/v1.0/debit/refund.htm",
        refund_order_request_obj,
        headers,
        400,  # Expected status code
        json_path_file,
        title_case,
        case_name,
        {"partnerReferenceNo": test_order_reference_number}
    )

@with_delay()
def test_refund_order_not_exist(test_order_reference_number):
    """Test refund when order not exist"""
    case_name = "RefundOrderInvalidBill"
    json_dict = get_request(json_path_file, title_case, case_name)
    
    json_dict["originalPartnerReferenceNo"] = "reference_number_not_exist"
    json_dict["partnerRefundNo"] = "reference_number_not_exist"
    json_dict["merchantId"] = merchant_id
    
    refund_order_request_obj = RefundOrderRequest.from_dict(json_dict)
    try:
        api_instance.refund_order(refund_order_request_obj)
        pytest.fail("Expected NotFoundException but the API call succeeded")
    except NotFoundException as e:
        assert_fail_response(json_path_file, title_case, case_name, e.body, {"partnerReferenceNo": json_dict["originalPartnerReferenceNo"]})
    except:
        pytest.fail("Expected NotFoundException but the API call give another exception")


@with_delay()
def test_refund_order_insufficient_funds(test_order_reference_number):
    """Test refund order insufficient funds"""
    case_name = "RefundOrderInsufficientFunds"

    # Get the request data from the JSON file
    json_dict = get_request(json_path_file, title_case, case_name)

    # Set the partner reference number
    json_dict["originalPartnerReferenceNo"] = test_order_reference_number
    json_dict["partnerRefundNo"] = test_order_reference_number
    json_dict["merchantId"] = merchant_id

    # Convert the request data to a RefundOrderRequest object
    refund_order_request_obj = RefundOrderRequest.from_dict(json_dict)

    # Make the API call and assert the response
    try:
        api_instance.refund_order(refund_order_request_obj)
        pytest.fail("Expected ForbiddenException but the API call succeeded")
    except ForbiddenException as e:
        assert_fail_response(json_path_file, title_case, case_name, e.body, {"partnerReferenceNo": json_dict["originalPartnerReferenceNo"]})
    except:
        pytest.fail("Expected ForbiddenException but the API call give another exception")

@with_delay()
def test_refund_order_unauthorized(test_order_reference_number):
    """Test refund order when authorization is invalid"""
    # Refund order
    case_name = "RefundOrderUnauthorized"

    # Get the request data from the JSON file
    json_dict = get_request(json_path_file, title_case, case_name)
    
    json_dict["originalPartnerReferenceNo"] = test_order_reference_number
    json_dict["partnerRefundNo"] = test_order_reference_number
    json_dict["merchantId"] = merchant_id

    # Convert the request data to a RefundOrderRequest object
    refund_order_request_obj = RefundOrderRequest.from_dict(json_dict)
    
    headers = get_headers_with_signature(
        method="POST",
        resource_path="/payment-gateway/v1.0/debit/refund.htm",
        request_obj=json_dict,
        with_timestamp=True,
        invalid_signature=True
    )
    
    execute_and_assert_api_error(
        api_client,
        "POST",
        "http://api.sandbox.dana.id/payment-gateway/v1.0/debit/refund.htm",
        refund_order_request_obj,
        headers,
        401,  # Expected status code
        json_path_file,
        title_case,
        case_name,
        {"partnerReferenceNo": test_order_reference_number}
    )

@with_delay()
def test_refund_order_merchant_status_abnormal(test_order_reference_number):
    """Test refund order merchant status abnormal"""
    case_name = "RefundOrderMerchantStatusAbnormal"

    # Get the request data from the JSON file
    json_dict = get_request(json_path_file, title_case, case_name)

    # Set the partner reference number
    json_dict["originalPartnerReferenceNo"] = test_order_reference_number
    json_dict["partnerRefundNo"] = test_order_reference_number
    json_dict["merchantId"] = merchant_id

    # Convert the request data to a RefundOrderRequest object
    refund_order_request_obj = RefundOrderRequest.from_dict(json_dict)

    # Make the API call and assert the response
    try:
        api_instance.refund_order(refund_order_request_obj)
        pytest.fail("Expected NotFoundException but the API call succeeded")
    except NotFoundException as e:
        assert_fail_response(json_path_file, title_case, case_name, e.body, {"partnerReferenceNo": json_dict["originalPartnerReferenceNo"]})
    except:
        pytest.fail("Expected NotFoundException but the API call give another exception")

@with_delay()
def test_refund_order_timeout(test_order_reference_number):
    """Test refund order timeout"""
    case_name = "RefundOrderTimeout"

    # Get the request data from the JSON file
    json_dict = get_request(json_path_file, title_case, case_name)

    # Set the partner reference number
    json_dict["originalPartnerReferenceNo"] = test_order_reference_number
    json_dict["partnerRefundNo"] = test_order_reference_number
    json_dict["merchantId"] = merchant_id

    # Convert the request data to a RefundOrderRequest object
    refund_order_request_obj = RefundOrderRequest.from_dict(json_dict)

    # Make the API call and assert the response
    try:
        api_instance.refund_order(refund_order_request_obj)
        pytest.fail("Expected ServiceException but the API call succeeded")
    except ServiceException as e:
        assert_fail_response(json_path_file, title_case, case_name, e.body, {"partnerReferenceNo": json_dict["originalPartnerReferenceNo"]})
    except:
        pytest.fail("Expected ServiceException but the API call give another exception")
        
@with_delay()
@retry_test(max_retries=3,delay_seconds=2)
def test_refund_order_idempotent():
    """Test refund order idempotent"""
    case_name = "RefundOrderIdempotent"
    
    # Create a test order paid and get the reference number
    order_reference_number = create_test_order_paid()

    # Get the request data from the JSON file
    json_dict = get_request(json_path_file, title_case, case_name)

    # Set the partner reference number
    json_dict["originalPartnerReferenceNo"] = order_reference_number
    json_dict["partnerRefundNo"] = order_reference_number
    json_dict["merchantId"] = merchant_id

    # Convert the request data to a RefundOrderRequest object
    refund_order_request_obj = RefundOrderRequest.from_dict(json_dict)
    
    # First hit API
    api_instance.refund_order(refund_order_request_obj)
    # Second hit API with same data
    api_response = api_instance.refund_order(refund_order_request_obj)
    # Assert the response
    assert_response(json_path_file, title_case, case_name, RefundOrderResponse.to_json(api_response), {"partnerReferenceNo": json_dict["originalPartnerReferenceNo"]})

File: test/java/src/test/java/id/dana/paymentgateway/RefundOrderTest.java
package id.dana.paymentgateway;

import id.dana.interceptor.CustomHeaderInterceptor;
import id.dana.invoker.Dana;
import id.dana.invoker.auth.DanaAuth;
import id.dana.invoker.model.DanaConfig;
import id.dana.invoker.model.constant.DanaHeader;
import id.dana.invoker.model.constant.EnvKey;
import id.dana.invoker.model.enumeration.DanaEnvironment;
import id.dana.paymentgateway.v1.api.PaymentGatewayApi;
import id.dana.paymentgateway.v1.model.*;
import id.dana.util.RetryTestUtil;
import id.dana.util.TestUtil;

import java.io.IOException;
import java.security.SecureRandom;
import java.util.*;
import java.util.concurrent.CountDownLatch;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;

import id.dana.util.ConfigUtil;
import okhttp3.OkHttpClient;
import org.junit.jupiter.api.Assertions;
import org.junit.jupiter.api.BeforeAll;
import org.junit.jupiter.api.Disabled;
import org.junit.jupiter.api.Test;

class RefundOrderTest {
    private static String jsonPathFile = RefundOrderTest.class.getResource("/request/components/PaymentGateway.json")
            .getPath();
    private static String userPin = "181818";
    private static String userPhone = "083811223355";
    private final String titleCase = "RefundOrder";
    private static final String merchantId = ConfigUtil.getConfig("MERCHANT_ID", "216620010016033632482");
    private static PaymentGatewayApi api;
    private static String
            partnerReferenceNoPaid,
            partnerReferenceNoInit;

    @BeforeAll
    static void setUp() {
        DanaConfig.Builder danaConfigBuilder = new DanaConfig.Builder();
        danaConfigBuilder
                .partnerId(ConfigUtil.getConfig("X_PARTNER_ID", ""))
                .privateKey(ConfigUtil.getConfig("PRIVATE_KEY", ""))
                .origin(ConfigUtil.getConfig("ORIGIN", ""))
                .env(DanaEnvironment.getByName(ConfigUtil.getConfig(EnvKey.ENV, "SANDBOX")));

        DanaConfig.getInstance(danaConfigBuilder);

        api = Dana.getInstance().getPaymentGatewayApi();

        List<String> dataOrder = createOrder();
        partnerReferenceNoInit = dataOrder.get(0);
    }

    @Test
    @RetryTestUtil.Retry
    void testRefundOrderValid() throws IOException {
        partnerReferenceNoPaid = payOrder(userPhone, userPin);
        String caseName = "RefundOrderValidScenario";
        RefundOrderRequest requestData = TestUtil.getRequest(jsonPathFile, titleCase, caseName,
            RefundOrderRequest.class);

        requestData.setOriginalPartnerReferenceNo(partnerReferenceNoPaid);
        requestData.setPartnerRefundNo(partnerReferenceNoPaid);
        requestData.setMerchantId(merchantId);

        Map<String, Object> variableDict = new HashMap<>();
        variableDict.put("partnerReferenceNo", partnerReferenceNoPaid);

        RefundOrderResponse response = api.refundOrder(requestData);
        TestUtil.assertResponse(jsonPathFile, titleCase, caseName, response, variableDict);
    }

    @Test
    void testRefundOrderInProgress() throws IOException {
        String caseName = "RefundOrderInProgress";
        Map<String, Object> variableDict = new HashMap<>();
        RefundOrderRequest requestData = TestUtil.getRequest(jsonPathFile, titleCase, caseName,
                RefundOrderRequest.class);

        requestData.setOriginalPartnerReferenceNo(partnerReferenceNoInit);
        requestData.setPartnerRefundNo(partnerReferenceNoInit);
        requestData.setMerchantId(merchantId);

        variableDict.put("partnerReferenceNo", partnerReferenceNoInit);
        RefundOrderResponse response = api.refundOrder(requestData);
        TestUtil.assertResponse(jsonPathFile, titleCase, caseName, response, variableDict);
    }

    @Test
    void testRefundOrderNotAllowed() throws IOException {
        String caseName = "RefundOrderNotAllowed";
        RefundOrderRequest requestData = TestUtil.getRequest(jsonPathFile, titleCase, caseName,
                RefundOrderRequest.class);
        requestData.setOriginalPartnerReferenceNo(partnerReferenceNoInit);
        requestData.setPartnerRefundNo(partnerReferenceNoInit);
        requestData.setMerchantId(merchantId);

        Map<String, Object> variableDict = new HashMap<>();
        variableDict.put("partnerReferenceNo", partnerReferenceNoInit);

        RefundOrderResponse response = api.refundOrder(requestData);
        TestUtil.assertResponse(jsonPathFile, titleCase, caseName, response, variableDict);
    }

    @Test
    void testRefundOrderNotPaid() throws IOException {
        String caseName = "RefundOrderNotPaid";
        RefundOrderRequest requestData = TestUtil.getRequest(jsonPathFile, titleCase, caseName,
                RefundOrderRequest.class);

        requestData.setOriginalPartnerReferenceNo(partnerReferenceNoInit);
        requestData.setPartnerRefundNo(partnerReferenceNoInit);
        requestData.setMerchantId(merchantId);

        Map<String, Object> variableDict = new HashMap<>();
        variableDict.put("partnerReferenceNo", partnerReferenceNoInit);

        RefundOrderResponse response = api.refundOrder(requestData);
        TestUtil.assertResponse(jsonPathFile, titleCase, caseName, response, variableDict);
    }

    @Test
    void testRefundOrderIllegalParameter() throws IOException {
        String caseName = "RefundOrderIllegalParameter";
        RefundOrderRequest requestData = TestUtil.getRequest(jsonPathFile, titleCase, caseName,
                RefundOrderRequest.class);
        requestData.setOriginalPartnerReferenceNo(partnerReferenceNoInit);
        requestData.setPartnerRefundNo(partnerReferenceNoInit);
        requestData.setMerchantId(merchantId);

        Map<String, Object> variableDict = new HashMap<>();
        variableDict.put("partnerReferenceNo", partnerReferenceNoInit);

        RefundOrderResponse response = api.refundOrder(requestData);
        TestUtil.assertResponse(jsonPathFile, titleCase, caseName, response, variableDict);
    }

    @Test
    void testRefundOrderInvalidMandatoryField() throws IOException {
        Map<String, String> customHeaders = new HashMap<>();
        String caseName = "RefundOrderInvalidMandatoryParameter";
        RefundOrderRequest requestData = TestUtil.getRequest(jsonPathFile, titleCase, caseName,
                RefundOrderRequest.class);

        requestData.setOriginalPartnerReferenceNo(partnerReferenceNoInit);
        requestData.setPartnerRefundNo(partnerReferenceNoInit);
        requestData.setMerchantId(merchantId);
        Map<String, Object> variableDict = new HashMap<>();
        variableDict.put("partnerReferenceNo", partnerReferenceNoInit);

        customHeaders.put(
                DanaHeader.X_SIGNATURE,
                "");
        OkHttpClient client = new OkHttpClient.Builder()
                .addInterceptor(new DanaAuth())
                .addInterceptor(new CustomHeaderInterceptor(customHeaders))
                .build();
        PaymentGatewayApi apiWithCustomHeader = new PaymentGatewayApi(client);

        RefundOrderResponse response = apiWithCustomHeader.refundOrder(requestData);
        TestUtil.assertResponse(jsonPathFile, titleCase, caseName, response, variableDict);
    }

    @Test
    void testRefundOrderNotExist() throws IOException {
        String orderNotExist = "f77466d6-1825-4091";
        String caseName = "RefundOrderInvalidBill";
        RefundOrderRequest requestData = TestUtil.getRequest(jsonPathFile, titleCase, caseName,
                RefundOrderRequest.class);
        requestData.setOriginalPartnerReferenceNo(orderNotExist);
        requestData.setPartnerRefundNo(orderNotExist);
        requestData.setMerchantId(merchantId);

        Map<String, Object> variableDict = new HashMap<>();
        variableDict.put("partnerReferenceNo", partnerReferenceNoInit);

        RefundOrderResponse response = api.refundOrder(requestData);
        TestUtil.assertResponse(jsonPathFile, titleCase, caseName, response, variableDict);
    }

    @Test
    void testRefundOrderInsufficientFunds() throws IOException {
        String caseName = "RefundOrderInsufficientFunds";
        RefundOrderRequest requestData = TestUtil.getRequest(jsonPathFile, titleCase, caseName,
                RefundOrderRequest.class);
        requestData.setOriginalPartnerReferenceNo(partnerReferenceNoInit);
        requestData.setPartnerRefundNo(partnerReferenceNoInit);
        requestData.setMerchantId(merchantId);

        Map<String, Object> variableDict = new HashMap<>();
        variableDict.put("partnerReferenceNo", partnerReferenceNoInit);

        RefundOrderResponse response = api.refundOrder(requestData);
        TestUtil.assertResponse(jsonPathFile, titleCase, caseName, response, variableDict);
    }

    @Test
    void testRefundOrderUnauthorized() throws IOException {
        Map<String, String> customHeaders = new HashMap<>();
        String caseName = "RefundOrderUnauthorized";

        RefundOrderRequest requestData = TestUtil.getRequest(jsonPathFile, titleCase, caseName,
                RefundOrderRequest.class);

        requestData.setOriginalPartnerReferenceNo(partnerReferenceNoInit);
        requestData.setPartnerRefundNo(partnerReferenceNoInit);
        requestData.setMerchantId(merchantId);

        customHeaders.put(
                DanaHeader.X_SIGNATURE,
                "85be817c55b2c135157c7e89f52499bf0c25ad6eeebe04a986e8c862561b19a5");
        OkHttpClient client = new OkHttpClient.Builder()
                .addInterceptor(new DanaAuth())
                .addInterceptor(new CustomHeaderInterceptor(customHeaders))
                .build();
        PaymentGatewayApi apiWithCustomHeader = new PaymentGatewayApi(client);

        RefundOrderResponse response = apiWithCustomHeader.refundOrder(requestData);
        TestUtil.assertResponse(jsonPathFile, titleCase, caseName, response, null);
    }

    @Test
    void testRefundOrderMerchantStatusAbnormal() throws IOException {
        String caseName = "RefundOrderMerchantStatusAbnormal";
        RefundOrderRequest requestData = TestUtil.getRequest(jsonPathFile, titleCase, caseName,
                RefundOrderRequest.class);
        requestData.setOriginalPartnerReferenceNo(partnerReferenceNoInit);
        requestData.setPartnerRefundNo(partnerReferenceNoInit);
        requestData.setMerchantId(merchantId);

        Map<String, Object> variableDict = new HashMap<>();
        variableDict.put("partnerReferenceNo", partnerReferenceNoInit);

        RefundOrderResponse response = api.refundOrder(requestData);
        TestUtil.assertResponse(jsonPathFile, titleCase, caseName, response, variableDict);
    }

    @Test
    void testRefundOrderTimeout() throws IOException {
        String caseName = "RefundOrderTimeout";
        RefundOrderRequest requestData = TestUtil.getRequest(jsonPathFile, titleCase, caseName,
                RefundOrderRequest.class);
        requestData.setOriginalPartnerReferenceNo(partnerReferenceNoInit);
        requestData.setPartnerRefundNo(partnerReferenceNoInit);
        requestData.setMerchantId(merchantId);

        Map<String, Object> variableDict = new HashMap<>();
        variableDict.put("partnerReferenceNo", partnerReferenceNoInit);

        RefundOrderResponse response = api.refundOrder(requestData);
        TestUtil.assertResponse(jsonPathFile, titleCase, caseName, response, variableDict);
    }

    @Test
    void testRefundOrderDuplicateRequest() throws IOException, InterruptedException {
        partnerReferenceNoPaid = payOrder(userPhone, userPin);
        String caseName = "RefundOrderDuplicateRequest";
        RefundOrderRequest requestData = TestUtil.getRequest(jsonPathFile, titleCase, caseName,
                RefundOrderRequest.class);
        String partnerReferenceNo = partnerReferenceNoPaid;
        requestData.setOriginalPartnerReferenceNo(partnerReferenceNo);
        requestData.setPartnerRefundNo(partnerReferenceNo);
        requestData.setMerchantId(merchantId);

        Map<String, Object> variableDict = new HashMap<>();
        variableDict.put("partnerReferenceNo", partnerReferenceNo);

        RefundOrderResponse response1 = api.refundOrder(requestData);
        Money money = new Money();
        money.setCurrency("IDR");
        money.setValue("10000.00");
        requestData.setRefundAmount(money);
        Thread.sleep(5000);
        RefundOrderResponse response2 = api.refundOrder(requestData);
        TestUtil.assertResponse(jsonPathFile, titleCase, caseName, response2, variableDict);
    }

    @Test
    void testRefundOrderIndempotent() throws InterruptedException {
        Map<String, Object> variableDict = new HashMap<>();
        String caseName = "RefundOrderIdempotent";
        int numberOfThreads = 10;
        ExecutorService executor = Executors.newFixedThreadPool(numberOfThreads);
        CountDownLatch latch = new CountDownLatch(numberOfThreads);

        RefundOrderRequest requestData = TestUtil.getRequest(jsonPathFile, titleCase, caseName,
                RefundOrderRequest.class);

        requestData.setOriginalPartnerReferenceNo(partnerReferenceNoInit);
        requestData.setPartnerRefundNo(partnerReferenceNoInit);
        requestData.setMerchantId(merchantId);
        variableDict.put("partnerReferenceNo", partnerReferenceNoInit);

        for (int i = 0; i < numberOfThreads; i++) {
            executor.submit(() -> {
                try {
                    RefundOrderResponse response = api.refundOrder(requestData);
                    TestUtil.assertResponse(jsonPathFile, titleCase, caseName, response, variableDict);
                    System.out.println("Thread: " + Thread.currentThread().getId()
                            + " - Status: " + response.getResponseCode());
                } catch (IOException e) {
                    throw new RuntimeException(e);
                } finally {
                    latch.countDown();
                }
            });
        }
        // Wait for all threads to complete
        latch.await();
        executor.shutdown();
    }

    public static List<String> createOrder() {
        List<String> dataOrder = new ArrayList<>();

        CreateOrderByApiRequest requestData = PaymentPGUtil.getCreateOrderApiRequest(
                jsonPathFile,
                "CreateOrder",
                "CreateOrderApi");

        // Assign unique reference and merchant ID
        String partnerReferenceNo = UUID.randomUUID().toString();
        requestData.setPartnerReferenceNo(partnerReferenceNo);
        requestData.setMerchantId(merchantId);
        requestData.setValidUpTo(PaymentPGUtil.generateDateWithOffset(30));

        Map<String, Object> variableDict = new HashMap<>();
        variableDict.put("partnerReferenceNo", partnerReferenceNo);

        CreateOrderResponse response = api.createOrder(requestData);
        Assertions.assertTrue(response.getResponseCode().contains("200"),
                "Response code should be 200, but was: " + response.getResponseCode());

        //Index 0 is partnerReferenceNo
        dataOrder.add(partnerReferenceNo);
        //Index 1 is the web redirect URL
        dataOrder.add(response.getWebRedirectUrl());

        return dataOrder;
    }

    public static String payOrder(String phoneNumber, String pin) {
        List<String> dataOrder = createOrder();
        PaymentPGUtil.payOrder(phoneNumber,pin,dataOrder.get(1));
        return dataOrder.get(0);
    }
}